packages/openai-compatible/src/openai-compatible-provider.ts

```diff
@@ -0,0 +1,188 @@
+import { OpenAICompatibleChatLanguageModel } from './openai-compatible-chat-language-model';
+import { OpenAICompatibleChatSettings } from './openai-compatible-chat-settings';
+import { OpenAICompatibleCompletionLanguageModel } from './openai-compatible-completion-language-model';
+import { OpenAICompatibleCompletionSettings } from './openai-compatible-completion-settings';
+import { OpenAICompatibleEmbeddingSettings } from './openai-compatible-embedding-settings';
+import { OpenAICompatibleEmbeddingModel } from './openai-compatible-embedding-model';
+
+export interface OpenAICompatibleProvider<
+  CHAT_MODEL_IDS extends string = string,
+  COMPLETION_MODEL_IDS extends string = string,
+  EMBEDDING_MODEL_IDS extends string = string,
+> extends ProviderV1 {
+  (
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  languageModel(
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  chatModel(
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+    options?: { defaultObjectGenerationMode: 'json' | 'tool' | undefined },
+  ): LanguageModelV1;
+
+  completionModel(
+    modelId: COMPLETION_MODEL_IDS,
+    settings?: OpenAICompatibleCompletionSettings,
+  ): LanguageModelV1;
+
+  textEmbeddingModel(
+    modelId: EMBEDDING_MODEL_IDS,
+    settings?: OpenAICompatibleEmbeddingSettings,
+  ): EmbeddingModelV1<string>;
+}
+
+export interface OpenAICompatibleProviderSettings {
+  /**
+Base URL for the API calls.
+     */
+  baseURL?: string;
+
+  /**
+API key for authenticating requests.
+     */
+  apiKey?: string;
+
+  /**
+Custom headers to include in the requests.
+     */
+  headers?: Record<string, string>;
+
+  /**
+Custom fetch implementation. You can use it as a middleware to intercept requests,
+or to provide a custom fetch implementation for e.g. testing.
+    */
+  fetch?: FetchFunction;
+
+  /**
+The name of the environment variable from which to load the API key (if a key isn't explicitly provided).
+   */
+  apiKeyEnvVarName?: string;
+
+  /**
+Description of the API key environment variable (for use in error messages).
+   */
+  apiKeyEnvVarDescription?: string;
+
+  /**
+Provider name. Overrides the `openai` default name for 3rd party providers.
+   */
+  name?: string;
+}
+
+/**
+Create an OpenAICompatible provider instance.
+ */
+export function createOpenAICompatible<
+  CHAT_MODEL_IDS extends string,
+  COMPLETION_MODEL_IDS extends string,
+  EMBEDDING_MODEL_IDS extends string,
+>(
+  options: OpenAICompatibleProviderSettings,
+): OpenAICompatibleProvider<
+  CHAT_MODEL_IDS,
+  COMPLETION_MODEL_IDS,
+  EMBEDDING_MODEL_IDS
+> {
+  if (!options.baseURL) {
+    throw new Error('Base URL is required');
+  }
+  const baseURL = withoutTrailingSlash(options.baseURL);
+
+  if (!options.name) {
+    throw new Error('Provider name is required');
+  }
+  const providerName = options.name;
+
+  const apiKey = loadApiKey({
```

Comment:
I would like to have these validation checks so that a dev is informed if they create a provider with unusable options. However, as I've done it here introduces an interesting failure case where just importing the file can cause these to throw due to the default provider pattern we use in concrete impls, e.g.:
- `generate-text/togetherai.ts` imports `createTogetherAI` or `togetherai` from `@ai-sdk/togetherai`
- the import happens before the `dotenv/config` import so the API key env var is not yet set
- the import of the togetherai provider logic causes the initialization of `const togetherai = createTogetherAI()` which will throw as there is no API key set
- a fix can include moving `dotenv/config` import before the TogetherAI imports

The above example is somewhat contrived and particular to our usage but if we hit it I am sure others would as well. Options include:
- do this validation as an early step in methods that would be impacted (but this would be annoying verbosity as there are a fair number of them, and it would be a small amount of redundant overhead on each call)
- only instantiate the default provider instance lazily (some add'l code complexity)
- don't do these checks (or at least, don't do the API key one)

Curious to hear your thoughts.

Reply:
as above, the easiest and potentially best way would be to not aim for extensibility / reuse for this particular file.

Reply:
example why: several providers will not have embeddings & completion models and they will need to limit the exposed API. There are many such cases and at the end just having a custom provider file for each one is imo easier to understand and maintain.